File: Windows/src/cloudprism/crypto/kdf.py
"""密钥派生函数封装。

使用 PBKDF2-HMAC-SHA256 从用户主密码 + 盐派生 32 字节 AES 密钥。
派生出的密钥仅存在于内存中，不落盘、不上传云端。

参数（SALT_LEN / ITERATIONS / KEY_LEN）必须与 Android 端完全一致，
否则同主密码在两端会派生出不同的密钥，导致互操作失败。
"""

from Crypto.Hash import SHA256, HMAC
from Crypto.Protocol.KDF import PBKDF2

from cloudprism import constants


class Kdf:
    """主密码派生对称密钥。

    所有方法均为静态，不持有任何状态；密钥由调用方临时持有并在用后清零。
    """

    # 盐值长度（字节）
    SALT_LEN: int = constants.SALT_LEN

    # PBKDF2 迭代次数
    ITERATIONS: int = constants.KDF_ITERATIONS

    # 派生密钥长度（字节），32 对应 AES-256
    KEY_LEN: int = constants.KEY_LEN

    # 算法标识
    ALGO: str = constants.KDF_ALGO

    @staticmethod
    def derive_key(master_password: str, salt: bytes) -> bytes:
        """从主密码 + 盐派生 32 字节对称密钥。

        参数:
            master_password: 用户主密码（明文，仅存内存）
            salt: KDF 盐值，长度建议为 SALT_LEN

        返回:
            32 字节派生密钥；调用方负责在使用后清零该内存
        """
        # 主密码统一 UTF-8 编码，保证 Windows 与 Android 字节级一致
        pw = master_password.encode("utf-8")

        # 显式指定 prf 为 HMAC-SHA256，避免不同平台 PBKDF2 默认 PRF 差异
        prf = lambda p, s: HMAC.new(p, s, SHA256).digest()

        return PBKDF2(
            password=pw,
            salt=salt,
            dkLen=Kdf.KEY_LEN,
            count=Kdf.ITERATIONS,
            prf=prf,
        )

File: WindowsGo/pkg/cryptox/kdf.go
package cryptox

import (
	"crypto/pbkdf2"
	"crypto/sha256"

	"github.com/Sagiri-lzumi/cloudprism/windowsgo/pkg/protocol"
)

// DeriveKey 从主密码 + 盐派生 32 字节 AES-256 密钥。
//
// 主密码统一按 UTF-8 编码后参与派生，保证 WindowsPy 与 Go 端字节级一致；
// 派生出的密钥仅存在于内存中，不落盘、不上传云端，调用方负责用后清零。
//
// 对照 WindowsPy/src/cloudprism/crypto/kdf.py:34-46
func DeriveKey(masterPassword string, salt []byte) [protocol.KeyLen]byte {
	return DeriveKeyRaw([]byte(masterPassword), salt)
}

// DeriveKeyRaw 从字节口令 + 盐派生 32 字节密钥，PBKDF2 参数与 DeriveKey 完全一致。
//
// 供非字符串密钥材料复用同一套参数 —— 目前唯一调用方是恢复码块：
// rkey = PBKDF2(恢复码随机密钥, rsalt)（见 vault.py:236）。
//
// 对照 WindowsPy/src/cloudprism/crypto/kdf.py:48-63
func DeriveKeyRaw(password, salt []byte) [protocol.KeyLen]byte {
	var key [protocol.KeyLen]byte

	// 使用 Go 1.24+ 标准库 crypto/pbkdf2 而非 golang.org/x/crypto/pbkdf2：
	// 零第三方依赖、由 FIPS 140 实现背书，且 HMAC-SHA256 走汇编加速。
	// 显式传 sha256.New，避免任何「默认 PRF」差异（Python 侧同样显式指定）。
	dk, err := pbkdf2.Key(sha256.New, string(password), salt, protocol.KDFIterations, protocol.KeyLen)
	if err != nil {
		// 迭代次数与密钥长度都是编译期常量，标准 PBKDF2 在此参数下不可能失败。
		// 唯一的失败路径是 GOFIPS140 模式下口令含 NUL 字节或盐短于 16 字节，
		// 二者都属于调用方违反协议契约（盐恒为 16 字节、口令为 UTF-8 文本），
		// 因此按编程错误处理，而不是给每个调用点都加一条不可达的错误分支。
		panic("cryptox: PBKDF2 参数非法: " + err.Error())
	}

	copy(key[:], dk)
	return key
}
